[ポーズデータの互換性について][英語版] 日本語版も少し修正 (#460)
vrm-c/vrm-specification: specification/VRMC_vrm_animation-1.0/how_to_transform_human_pose.md

# About pose data compatibility

> > Non-normative

Reviewer: ```suggestion
> This document is non-normative.
```

## Terms

### Humanoid
*Humanoid* referred to in this document depends on the VRM Humanoid definition.

[Humanoid definition within the VRM specification](https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0/humanoid.ja.md)

### T-pose
*T-pose* referred to in this document depends on T-pose of VRM.
[VRM T-pose: Specification of posture defined by VRM](https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0/tpose.ja.md)

### rest rotation
The rotation state of node in T-pose state defined by VRM-1.0 is **Rest rotation**.

VRM-0.X made pose data compatible by forcing rest rotations to be no rotations.
VRM-1.0 eliminates this no-rotation restriction on rest rotation.

### Non-required bone
**Non-required bones** in VRM may have different presence or absence between models.

For example, bones such as `hips` and `head` are required bones, whereas bones such as `upperChest` and `leftShoulder` are non-required bones.
Therefore, there are models with and without the `upperChest` bone.

### same posture
Suppose each `HumanBone` has the **same pose** when it is visually oriented the same way.

> Subtle differences such as being slightly slanted during T-pose are included. It won't be in exactly the same position.

Reviewer: ```suggestion
Subtle differences, such as being slightly slanted during T-pose, are considered as the same posture so that it won't be in exactly the same posture.

```

### pose data
**Pose data** is represented by translation values of the `hips` bone and local rotations of all Humanoid bones.

The values for the same posture will change under the influence of rest rotation.
Also, the values for the same posture will change depending on the presence or absence of non-required bones.

> **If the rest rotation of T-pose or the presence or absence of non-essential bones are different, pose data for the same pose will not be compatible.**

Reviewer: ここは、"will not be compatible" （原文では「互換性がありません」）と言うよりは、その直後の節との繋がりも意識し、「そのまま扱うことはできません」（"cannot be used as-is"）というニュアンスにしたほうが良い気がしてきました。

Reviewer: done @ https://github.com/vrm-c/vrm-specification/pull/460/commits/de585c3f16b3bef014bbd1a22df0d6a14689ad9f

## Conversion processing to make pose data compatible

In order to make pose data compatible between models with different T-poses, one of the following conversion processes is required.

### Convert the model
Converts the rest rotation of the model to no rotation.

In VRM-0.X (UniVRM), the rest rotation of the model was converted to no rotation when exporting the model.

### Convert pose data
Convert pose data to fit a model with arbitrary rest rotation.

In VRM-1.0, this method is assumed because the model has arbitrary rest rotation.

VRM-1.0 (UniVRM) provides a function called **ControlRig** that allows pose data conversion at runtime.

## Conversion process details

- `TPoseA`: Model A rest rotation
- `PoseForA`: Pose data that gives the intended look when applied to model A
- `TPoseB`: Model B rest rotation

> Given the above, we will explain how to obtain pose data that looks the same as `PoseForA` when applied to `PoseForB`: Model B.

Reviewer: ```suggestion
Given the above, we will explain how to obtain pose data `PoseForB` that looks the same as `PoseForA` when applied to Model B.

```

### intermediate format: NormalizedLocalRotation

To simplify processing here, we introduce an intermediate format: `NormalizedLocalRotation`.

`PoseForA` => `NormalizedLocalRotation` => `PoseForB`

`NormalizedLocalRotation` is pose data that looks the same as `PoseForA` when applied to a model with no rest rotation.

> The intermediate format is almost the same as the normalized state of VRM-0.X. The definition of T-pose was clarified in VRM-1.0, so it cannot be said that they are completely the same.
>

> > If `TPoseA` has been normalized, `PoseForA` and `NormalizedLocalRotation` will be equal, making it easier.

Reviewer: @ousttrue  英語版OKです。原文「正規化」「無回転」の変更もついでにやっちゃって大丈夫です！

> > If `TPoseA` has been normalized, `PoseForA` and `NormalizedLocalRotation` will be equal, making it easier.

Author: 日本語版を追随しました 🙏

> > Similarly, if `TPoseB` has been normalized, `NormalizedLocalRotation` and `PoseForB` will be equal, making it easier.

Reviewer: ```suggestion
> Similarly, if `TPoseB` has no rotations, `NormalizedLocalRotation` and `PoseForB` will be equal, making it easier.
```

### `PoseForA` => `NormalizedLocalRotation`

- W: World rest rotation of TPoseA
- L: Local rest rotation of TPoseA

$NormalizedLocalRotation = W \cdot L^{-1} \cdot A.LocalRotation \cdot W^{-1}$

```cs
// C#
InitialGlobalRotation * Quaternion.Inverse(InitialLocalRotation) * Transform.localRotation * Quaternion.Inverse(InitialGlobalRotation);
```

### `NormalizedLocalRotation` => `PoseForB`

> - W: Local rest rotation of TPoseA

Reviewer: ```suggestion
- W: World rest rotation of TPoseB
```

- L: Local rest rotation of TPoseB

$B.LocalRotation = L \cdot W^{-1} \cdot NormalizedLocalRotation \cdot W$

```cs
// C#
ControlTarget.localRotation = _initialTargetLocalRotation * (Quaternion.Inverse(_initialTargetGlobalRotation) * ControlBone.localRotation * _initialTargetGlobalRotation);
```

### When the presence or absence of non-required bones differs
This section explains the conversion process when the presence or absence of non-required bones differs between the conversion source model and conversion destination model.

#### When the source model has few bones
If the source model has few bones, it is recommended to simply apply only the bones contained in the source model.

#### When the target model has few bones
If the destination model has few bones, it is recommended to apply the rotation of the target bone to all child bones that would otherwise be affected by the rotation of the target bone.

> For example, if upperChest exists in the source model but not in the destination model, the neck, leftShoulder, and rightShoulder of the destination model will be applied by multiplying both the rotation of upperChest and the rotation of the bone itself. It is desirable that

Reviewer: ```suggestion
For example, if upperChest exists in the source model but not in the destination model, the neck, leftShoulder, and rightShoulder of the destination model should be transformed by both the rotation of upperChest and the rotation of the bone itself by multiplying the two.
```
